clang/lib/CodeGen/CodeGenAction.cpp

```diff
@@ -1034,14 +1038,25 @@ CodeGenAction::CreateASTConsumer(CompilerInstance &CI, StringRef InFile) {
     CI.getPreprocessor().addPPCallbacks(std::move(Callbacks));
   }
 
+  std::vector<std::unique_ptr<ASTConsumer>> AdditionalConsumers;
+  AdditionalConsumers.reserve(2);
+
   if (CI.getFrontendOpts().GenReducedBMI &&
       !CI.getFrontendOpts().ModuleOutputPath.empty()) {
-    std::vector<std::unique_ptr<ASTConsumer>> Consumers(2);
-    Consumers[0] = std::make_unique<ReducedBMIGenerator>(
+
+    AdditionalConsumers.push_back(std::make_unique<ReducedBMIGenerator>(
         CI.getPreprocessor(), CI.getModuleCache(),
-        CI.getFrontendOpts().ModuleOutputPath);
-    Consumers[1] = std::move(Result);
-    return std::make_unique<MultiplexConsumer>(std::move(Consumers));
+        CI.getFrontendOpts().ModuleOutputPath));
+  }
+
+#if CLANG_ENABLE_CIR
+  if (CI.getFrontendOpts().ClangIRAnalysisOnly)
+    AdditionalConsumers.push_back(cir::createCIRAnalysisOnlyConsumer(CI));
```

Comment:
I'm a bit uncomfortable with touching anything under `clang/lib/CodeGen`, this is going to be a hassle to upstream / convince maintainers. Can't this be done instead in the opposite way? In CIRGenAction.cpp we could dispatch the classic CodeGen if the pipeline is asking for LLVM codegen and `-fclangir-analysis-only` is ON. Have you tried that and found any issue?

Reply:
> I'm a bit uncomfortable with touching anything under clang/lib/CodeGen

This is the style we did for modules. It shows it is pretty stable.

> Can't this be done instead in the opposite way? 

But if you really want, we can do it. We just need to introduce more lines of code. Please see the newest patch.

Reply:
> This is the style we did for modules. It shows it is pretty stable.

The approach is sounds, my point is more about trying to minimize clang changes outside CIR dirs, so for the time being is less invasive to the rest of the codebase,

> But if you really want, we can do it. We just need to introduce more lines of code. Please see the newest commit.

This looks great, thanks!